cmd/gateway-gcs.go

```diff
@@ -601,18 +659,32 @@ func fromGCSMultipartKey(s string) (key, uploadID string, partID int, err error)
 	return
 }
 
-func toGCSMultipartKey(key string, uploadID string, partID int) string {
-	// we don't use the etag within the key, because the amazon spec
-	// explicitly notes that uploaded parts with same number are being overwritten
+func unescape(s string) string {
+	s = strings.Replace(s, "%2D", "-", -1)
+	s = strings.Replace(s, "%2F", "/", -1)
+	s = strings.Replace(s, "%25", "%", -1)
+	return s
+}
+
+func escape(s string) string {
+	s = strings.Replace(s, "%", "%25", -1)
+	s = strings.Replace(s, "/", "%2F", -1)
+	s = strings.Replace(s, "-", "%2D", -1)
+	return s
+}
 
+func toGCSMultipartKey(key string, uploadID string, partID int) string {
 	// parts are allowed to be numbered from 1 to 10,000 (inclusive)
-	return fmt.Sprintf("%s/multipart-%s-%s-%05d", ZZZZMinioPrefix, key, uploadID, partID)
+
+	// we need to encode the key because of possible slashes
+	return fmt.Sprintf("%s/multipart-%s-%s-%05d", ZZZZMinioPrefix, escape(key), uploadID, partID)
 }
 
 // NewMultipartUpload - upload object in multiple parts
 func (l *gcsGateway) NewMultipartUpload(bucket string, key string, metadata map[string]string) (uploadID string, err error) {
 	// generate new uploadid
 	uploadID = mustGetUUID()
+	uploadID = strings.Replace(uploadID, "-", "", -1)
```

Comment:
You can put this into a function and get a different UUID representation. @nl5887